pkg/globalnet/controllers/ipam/gatewaymonitor.go

```diff
@@ -124,197 +127,104 @@ func (gm *GatewayMonitor) Stop() {
-			gm.endpointWorkqueue.Forget(obj)
-			return fmt.Errorf("error retrieving submariner endpoint list %v", err)
+			// Ideally this case will never hit, as the subnets are valid CIDRs
+			klog.Warningf("unable to validate overlapping Service CIDR: %s", err)
 		}
 
-		for _, endpoint := range endpoints.Items {
-			if endpoint.Spec.ClusterID != gm.clusterID {
-				for _, remoteSubnet := range endpoint.Spec.Subnets {
-					MarkRemoteClusterTraffic(gm.ipt, remoteSubnet, AddRules)
-				}
-			}
-		}
+		if overlap {
+			// When GlobalNet is used, globalCIDRs allocated to the clusters should not overlap.
+			// If they overlap, skip the endpoint as its an invalid configuration which is not supported.
+			klog.Errorf("GlobalCIDR %q of local cluster %q overlaps with remote cluster %s",
+				gm.ipamSpec.GlobalCIDR[0], gm.ipamSpec.ClusterID, endpoint.Spec.ClusterID)
 
-		hostname, err := os.Hostname()
-		if err != nil {
-			klog.Fatalf("Unable to determine hostname: %v", err)
+			return false
 		}
 
-		// If the endpoint hostname matches with our hostname, it implies we are on gateway node
-		if endpoint.Spec.Hostname == hostname {
-			klog.V(log.DEBUG).Infof("We are now on GatewayNode %s", endpoint.Spec.PrivateIP)
-
-			configureTCPMTUProbe()
-
-			gm.syncMutex.Lock()
-			if !gm.isGatewayNode {
-				gm.isGatewayNode = true
-				gm.initializeIPAMController(gm.ipamSpec.GlobalCIDR[0], gm.nodeName)
-			}
-			gm.syncMutex.Unlock()
-		} else {
-			klog.V(log.DEBUG).Infof("We are on non-gatewayNode. GatewayNode ip is %s", endpoint.Spec.PrivateIP)
-
-			gm.syncMutex.Lock()
-			if gm.isGatewayNode {
-				gm.stopIPAMController()
-				gm.isGatewayNode = false
-			}
-			gm.syncMutex.Unlock()
+		for _, remoteSubnet := range endpoint.Spec.Subnets {
+			MarkRemoteClusterTraffic(gm.ipt, remoteSubnet, AddRules)
 		}
 
-		gm.endpointWorkqueue.Forget(obj)
-
-		return nil
-	}()
+		return false
+	}
 
+	hostname, err := os.Hostname()
 	if err != nil {
-		utilruntime.HandleError(err)
-		return true
+		klog.Fatalf("Unable to determine hostname: %v", err)
 	}
 
-	return true
-}
+	// If the endpoint hostname matches with our hostname, it implies we are on gateway node
+	if endpoint.Spec.Hostname == hostname {
+		klog.V(log.DEBUG).Infof("We are now on GatewayNode %s", endpoint.Spec.PrivateIP)
 
-func (gm *GatewayMonitor) enqueueEndpoint(obj interface{}) {
-	var key string
-	var err error
-	if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
-		utilruntime.HandleError(err)
-		return
-	}
+		configureTCPMTUProbe()
 
-	klog.V(log.TRACE).Infof("Enqueueing endpoint in gatewayMonitor %v", obj)
-	gm.endpointWorkqueue.AddRateLimited(key)
-}
-
-func (gm *GatewayMonitor) handleRemovedEndpoint(obj interface{}) {
-	var object *v1.Endpoint
-	var ok bool
-	if object, ok = obj.(*v1.Endpoint); !ok {
-		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
-		if !ok {
-			klog.Errorf("Could not convert object %v to an Endpoint", obj)
-			return
+		gm.syncMutex.Lock()
+		if !gm.isGatewayNode {
```

Comment:
This part of code gets simplified if we move to Event handler interface, but that can be done in a separate PR.